Railswitch overhaul

SirAlador opened the pull request:
I changed a lot of things about RailSwitch. Lets go:

## New Features
- RailSwitch now checks for valid switch signs on all four sides of the rail (north, south, east, west) in addition to the block above the rail. The block above the rail is checked first, so no existing setups should break.
- Destination matching with regular expressions! `[destex]` switches concatenate all 3 lines into one regex, and try to match it against each of the user's destinations. See the [README](https://github.com/SirAlador/RailSwitch/blob/master/README.md#regularexpression-destination-check) for usage. Regular expressions are matched using the [REJ2](https://github.com/google/re2j) library, which takes out some features to match regular expressions in linear time. This means that players cannot troll the server with [catastrophic backtracking](https://www.regular-expressions.info/catastrophic.html). 
- `[destadd]`, `[destrm]`, and `[destrmex]`, switches to automatically set and unset destinations from the player. See the [README](https://github.com/SirAlador/RailSwitch/blob/master/README.md#destination-adding) for usage.
- If any exception is thrown during switch logic, a smoke particle will emit from the switch sign. This is useful for knowing when you made a typo in your regex.

## Technical changes
- Updated to CivGradle 2.0
- Added a dependency on the REJ2 library. The jar must now be built with `build shadowJar reobf`.
- Changed the dependencies on CivModCore, Namelayer, and Citadel to `compileOnly` as a result of the above.
- Logic for finding switch signs in `SwitchListener` is now in a for loop, and a bunch of variables were moved/renamed.
- The `SwitchType` enum was replaced with the `SwitchLogic` abstract class, and logic for choosing a switch type is now in `SwitchLogic.try_create`.
- The actual switch logic was moved from `SwitchListener` to `SimpleDestLogic`.
- The destination setting now has an enforced limit of 256 characters.

## Food for thought
- Should the dependency on REJ2 be moved to CivModCore, since it already has 3rd-party dependencies? Would be useful if any other plugins wanted to use it, but it would also be a little awkward to just have it sitting there.
- When I first asked SoundTech about this idea many months ago, he said it would be awesome if I could make things configurable. I decided to not do this, as if the server begins to depend upon these new features, disabling them in config would be akin to just disabling the entire RailSwitch plugin. Thoughts?
- Should I commit the test world into this repository? As a folder or as a zip?

## Testing
I have tested this locally in isolation only with CivModCore and no database. I have not tested it with full CivGradle, as I imagine my laptop would combust running CivGradle and Minecraft at the same time.

Here is my test setup: [https://www.dropbox.com/s/2k3v67d4ndi32bg/railswitch_test.zip](https://www.dropbox.com/s/2k3v67d4ndi32bg/railswitch_test.zip). There's one long test track that goes through about every test I could think of, and then a worst-scenario test to make sure catastrophic backtracking is actually not a problem. The test world requires that command blocks are enabled.

Super excited to finally get this out the door! Tell me your comments, critiques, and/or concerns.

SirAlador committed:
🪧 Support multiple sign locations

- Select the first rail sign from an array of locations
- Added the block below the block that the rail rests on, in addition to
  the block above the rail, to the locations checked

SirAlador committed:
📜 Abstract switch logic, add RegEx switch logic

SirAlador committed:
😅 Fix small things

SirAlador committed:
☁️ Emit a smoke particle when switch logic fails

SirAlador committed:
🦺 Switch to RE2J, added [destadd] and [destrm]

SirAlador committed:
🤪 Actually commit the changes to gradle files

SirAlador committed:
🐞 Bug fixes

SirAlador committed:
🤦🏻‍ Re-add [destination] and [!destination]

SirAlador committed:
📓 Update README

SirAlador committed:
🤓 Don't mix spaces and tabs

SirAlador committed:
✏️ Fix spelling in README

SirAlador committed:
🔥 Deleted obsolete SwitchType.java

SirAlador committed:
✏️ More README fixes

Protonull commented:
> Should the dependency on REJ2 be moved to CivModCore, since it already has 3rd-party dependencies?

If this is something we're concerned about, perhaps we should start using that [`libraries`](https://docs.papermc.io/paper/dev/plugin-yml#libraries) feature.

Protonull commented on paper/src/main/java/sh/okx/railswitch/switches/SwitchListener.java:
```diff
@@ -84,46 +117,26 @@ public void onSwitchTrigger(BlockRedstoneEvent event) {
+        String dest_string = SettingsManager.getDestination(player);
+        try {
+            if (logic.decide(player)) event.setNewCurrent(15);
+            else event.setNewCurrent(0);
```
if you're trying to make this as compact as possible, just turn this into a ternary:
```java
event.setNewCurrent(logic.decide(player) ? 15 : 0);
```

Protonull commented on paper/src/main/java/sh/okx/railswitch/switches/SwitchListener.java:
```diff
@@ -84,46 +117,26 @@ public void onSwitchTrigger(BlockRedstoneEvent event) {
     }
 
+    /** Emit a campfire particle at the location of the sign to indicate failure */
+    public void emit_failure_smoke(Location block_coordinates) {
+        Location location = block_coordinates.add(0.5, 0.5, 0.5);
+        location.getWorld().spawnParticle(Particle.CAMPFIRE_COSY_SMOKE, location, 0, 0.0, 0.0, 0.0, 3.0);
+    }
```
Setting aside the snake case for a moment, you can readily one-liner this with:
```java
public void emit_failure_smoke(Location block_coordinates) {
    block_coordinates.getWorld().spawnParticle(block_coordinates.toCenterLocation(), 0, 0.0, 0.0, 0.0, 3.0);
}
```
[API Reference](https://jd.papermc.io/paper/1.18/org/bukkit/Location.html#toCenterLocation())

Protonull commented on paper/src/main/java/sh/okx/railswitch/switches/DestExUtils.java:
```diff
@@ -0,0 +1,29 @@
+
+public class DestExUtils {
+	static Pattern compile_pattern(String[] lines, String raw_flags) throws Exception {
+		StringBuilder pattern_appender = new StringBuilder();
+		for (String line : lines) pattern_appender.append(line);
+		String pattern_text = pattern_appender.toString();
```
Could be wrong, but couldn't you use `String.join("", lines)` here?

Protonull commented on paper/src/main/java/sh/okx/railswitch/switches/RemoveDestExLogic.java:
```diff
@@ -0,0 +1,51 @@
+		Iterator<String> iter = dests.iterator();
+		while (iter.hasNext()) {
+			String dest = iter.next();
+			if (dest.trim().length() == 0) continue;
```
Use [`.isBlank()`](https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/lang/String.html#isBlank()) instead

Protonull commented on paper/src/main/java/sh/okx/railswitch/switches/RemoveDestExLogic.java:
```diff
@@ -0,0 +1,51 @@
+	@Override
+	public boolean decide(Player player) {
+		boolean did_something = false;
+		ArrayList<String> dests = new ArrayList(Arrays.asList(SettingsManager.getDestination(player).split(" ")));
```
As a general rule, [don't use String.split()](https://stackoverflow.com/a/44401073)

SirAlador replied:
Dang, that's interesting. Did not know.

Protonull commented on paper/src/main/java/sh/okx/railswitch/switches/SwitchLogic.java:
```diff
@@ -0,0 +1,43 @@
+		if (destexrm.matches()) return new RemoveDestExLogic(lines, destexrm);
+
+		//Failure to specify the [] bits
+		if (lines[0].length() >= 2 && lines[0].charAt(0) == '[' && lines[0].charAt(lines[0].length() - 1) == ']') {
```
Can be replaced with [`lines[0].startsWith("[")`](https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/lang/String.html#startsWith(java.lang.String)) and [`lines[0].endsWidth("]")`](https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/lang/String.html#endsWith(java.lang.String))

SirAlador committed:
✅ Implement Protonull's suggestions

https://github.com/CivMC/RailSwitch/pull/9

SirAlador committed:
🛤 Merge branch 'master' of github.com:SirAlador/RailSwitch

Diet-Cola commented:
Very cool, dont forget to credit yourself by adding yourself to the authors list in plugin.yml 😄

Protonull commented:
DestIterator seems a bit overkill for the problem. You could've simply replaced `String.split()` with [`StringUtils.split()`](https://commons.apache.org/proper/commons-lang/apidocs/org/apache/commons/lang3/StringUtils.html#split-java.lang.String-java.lang.String-), which *should* be available to you via PaperMC or CivModCore. The only other issue, which is more an aesthetic issue, is the prolific use of snake\_case for method, variable, and field names ([reference](https://www.javatpoint.com/java-naming-conventions))

SirAlador commented:
> You could've simply replaced `String.split()` with [`StringUtils.split()`](https://commons.apache.org/proper/commons-lang/apidocs/org/apache/commons/lang3/StringUtils.html#split-java.lang.String-java.lang.String-)

Ah, didn't know we had Apache commons on the classpath already

> prolific use of snake_case for method, variable, and field names

awwwwww. Will adjust.

SirAlador committed:
🐫 reject_modernity, returnToCamel; StrinUtils.split()

- switch from snake_case to camelCase
- replace uses of DestIterator with apache common's StringUtils.split
- rename classes to be consistent with their invocations

SirAlador committed:
🎩 Attribute myself

SirAlador commented:
I'm now realizing that the diff for [this commit](https://github.com/CivMC/RailSwitch/pull/9/commits/2f8d525dc603d0302e3c432c9782070376b696b5) is really sloppy -- didn't think too much about how renaming and changing files in the same commit would work. If you'd like me to redo it as two commits, let me know.

okx-code commented:
I think that supporting regexes adds a bit too much complexity, makes compilation harder, and affects maintainability. I think the regexes can be simplified if they only supported `*`, which would still allow many use cases. This would also involve removing regular expression flags which would then help to simplify other parts of the plugin, and make regex matching the first line not necessary. Currently, in all cases RailSwitch is case insensitive and I don't see why that needs to be possible to disable.

I also don't agree with blanket catching Exception in many cases, I think that can be improved significantly by only using exceptions for what is absolutely necessary, and letting unknown cases fail rather than attempting to proactively catch something we don't know about.
For example, the sign is not valid in tryCreate it should return null, rather than throwing an exception.

okx-code commented:
If you are still interested in this refactor, please open it in https://github.com/CivMC/Civ
